topdown/object.go

```diff
@@ -88,6 +90,63 @@ func builtinObjectGet(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Ter
 	return iter(operands[2])
 }
 
+func builtinObjectDig(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// top level object input must type check as an Object
+	object, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// paths must type check as Array
+	path, err := builtins.ArrayOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	// if the path is empty, then we skip digging and return the default
+	if path.Len() == 0 {
+		return iter(operands[2])
+	}
+
+	var dugTerm *ast.Term
+
+	// index is used to track if the term being processed is the last in the path
+	var index int
+	err = path.Iter(func(term *ast.Term) error {
+		index++
+
+		// dugTerm will either be the return value, or the object to dig for the next iteration
+		dugTerm = object.Get(term)
+		if dugTerm == nil {
+			return fmt.Errorf("dugTerm not found")
```

Comment:
These won't make much sense to a user who sees them (via `opa eval --strict-builtin-errors` for example). Perhaps we should say something like "nested term is not an object" and give some indication on where that failed...?

Ah they'll never surface because of the default value. Hrm. Does this make it hard to debug? I'm not a fan of default args... but it does match how `object.get` behaves... 🤔

Reply:
> Ah they'll never surface because of the default value

Yeah, I was quite uncertain about what a good way to do this was. I was keen to make it similar to object.get. I also need to return errors if using this implementation / Iter. I couldn't see an easy way to `for ... range` over the array.

Reply:
Perhaps if I just had a generic `digError` and return that in the case of any error? Idk, open to ideas from others about what's best.

Guess it comes down to what people think about the defaulting behaviour.